### ISSUE TITLE
Filter by District

### ISSUE THREAD

**Issue by technologynhaa** [code4recovery/12-step-meeting-list#189]
One of our districts went through the process of comparing the current TSML Meeting list to the Districts Website, The Area Website and the Meeting App. The idea being to assure all elements were the same and the Zoom links functioned at all levels.

In TSML, if a by District filter could be added to the online meeting list, it would ease searching for District specific meetings.

**Comment by tech2serve** [comment]
What do you have here?

<img width="181" alt="image" src="https://user-images.githubusercontent.com/31904316/96377081-929cd000-1148-11eb-8fdb-b2d285b0065d.png">

**Comment by technologynhaa** [comment]
We have a list of towns in our Area.

**Comment by tech2serve** [comment]
Oops. I edited my comment, probably the same time you were replying to me.

Do you have your meetings assigned to their respective group? And is that group assigned to a district?

**Comment by technologynhaa** [comment]
We do in our Area Database.  How would I determine that from a TSML
perspective?

**Comment by tech2serve** [comment]
On the Admin side, open a meeting for editing and scroll down to the contact information. If your meetings are assigned to groups and districts, you should see something like this:

<img width="945" alt="image" src="https://user-images.githubusercontent.com/31904316/96378344-edd2c080-1150-11eb-982e-50113da21cd6.png">

Then, you should see "Switch to Districts" show up under the Locations filter. Since you don't have that, I'm guessing your database has implemented Districts. If I'm wrong, please send me a screen cap showing one of them, and we'll troubleshoot why it isn't showing up under the Locations filter.

**Comment by tech2serve** [comment]
Going to close this for now. If the OP returns, we can reopen as necessary.

**Comment by technologynhaa** [comment]
I need to figure out how to get the data in

**Comment by technologynhaa** [comment]
Attached is the csv format we use.  Is it not in the correct format?  It is
not populating District.  I see your comment that District ought to be a
string, but just want to understand exactly what we need to do.  Are you
able to reopen this ticket?

**Comment by tech2serve** [comment]
Re-opened.

However, the csv did not come through. Please try to attach it again.

**Comment by stefanklehmann** [comment]
We're having the same issue at Area 79.  We're batch importing over 700 meetings from our old website, and almost none of them have a group associated with them (in the current data set).  It sounds from the above like we have to first assign each meeting to a group before TSML will allow us to associate a meeting with a District.  Is that right @tech2serve?

**Comment by tech2serve** [comment]
I confess I haven't tested this with the csv importer, but if nothing has changed you should be able to include group and district information in the csv and import it. Groups (providing service to/through one or more meetings) and districts follow the AA service structure which is why they're related in TSML.

If the district information is present, the filter by district option is present and should work. 

My suggestion to @stefanklehmann and @technologynhaa is to add the information to your CSV or to your database through the edit meeting page, and try again. The `template.csv` provided as an example in the plugin shows the  I don't think there is anything regarding this feature that doesn't appear to be working. 

Please report back.

**Comment by technologynhaa** [comment]
Happy Thanksgiving!  I attempted to copy the data in the csv I originally
sent to the format of your template, but when I import, it does not produce
any results.  What shall I do?  If I can get a version of the template that
imports correctly, I can get the exporter of the template to conform to
this format and the it will work on an ongoing basis.

Yours In Service,

Sean K.
Area 43 Technology Chair

**Comment by tim-rohrer** [comment]
I frequently use the `template.csv` to reset my dev site. What spreadsheet program are you using?

**Comment by technologynhaa** [comment]
Numbers

**Comment by technologynhaa** [comment]
The strange thing is even exporting from TSML and Re importing does not
work.

**Comment by tech2serve** [comment]
Any errors in your browser console? Please explain more of what you mean by "doesn't work" Try disabling other plugins. Etc., etc. I believe this to be something unique to your site.

**Comment by technologynhaa** [comment]
I have attached a screenshot of the console.  What I mean is that I can use
the attached meetingdata.csv which is produced by areafellowship.com (the
database we use) and it imports correctly.  When I save a copy of our
meetings from the plugin to the attached meetings.csv produced by TSML and
attempt to re import, it does not work.   I have disabled all plugins and
it still does not work.  There may be something unique to our site, but I
am unable to find it.  Would it be possible to do a screen share or would
you be able to attempt to reproduce?

Yours in Service,

Sean K.
Area 43 Technology Chair

**Comment by tech2serve** [comment]
@technologynhaa I think something may be wrong with how you're attaching things??? I never got your original csv, and now I don't see any screen cap. 

I'm happy to try and duplicate if you can get me the csv. We'll then go from there.

**Comment by technologynhaa** [comment]
Attachments do not seem to go through e-mail.  I am putting them in github directly.  It does not support csv attachments, so I had to compress.
[meetingdata.csv.zip](https://github.com/code4recovery/12-step-meeting-list/files/5611179/meetingdata.csv.zip)
![Screen Shot 2020-11-28 at 9 17 46 AM](https://user-images.githubusercontent.com/73075721/100525266-2c668c80-318d-11eb-85df-578739894496.png)


[meetings.csv.zip](https://github.com/code4recovery/12-step-meeting-list/files/5611178/meetings.csv.zip)

**Comment by tech2serve** [comment]
Hey @technologynhaa --

I took your meetings.csv file and imported it without any warnings or errors. 

![image](https://user-images.githubusercontent.com/31904316/100528275-9a687f00-31a0-11eb-8300-fdfbd02f9385.png)

I must believe then that something is interfering with the process on your site, and would suggest you start with the common troubleshooting steps:

1. Deactivate all other plugins (except Enable jQuery Migrate Helper which you should have installed if it isn't) and see if the problem resolves.
1. Try a virgin browser, meaning one without any extensions or other added coded that might mess with the process.
1. Clear caches for your browser and staging site.
1. Confirm your site is using a supported version of php.

**Comment by technologynhaa** [comment]
Hey,

I followed all of the troubleshooting steps you indicated and got the same
result.  The version of PHP is  7.3.24 (Supports 64bit values).  Just out
of curiosity, did you try uploading meetingsdata.csv?  Also, where would
the tsml_import_reformat be located?  Perhaps there was some customization
that was implemented that is causing this behavior.

Thanks in advance,

Sean K.
Area 43 Technology Chair

**Comment by tech2serve** [comment]
I did not try to import meetingsdata.csv as there were no headers and I was reasonably certain that data was from your other database, and would not be compatible directly.

You have to create the `tsml_import_reformat` function yourself. Typically, the best approach for this is to create a child theme and then add your own `functions.php`. Search the Internet for steps on how to do with, or ask others in TIAA Forum for help. Once that is set up, you then need to develop a reformatting function and add that to your own `functions.php`. This approach prevents your customizations from being overwritten when the plugin is updated.

But if you cannot import your meetings.csv file because of the non-unique nonce issue, then something is likely wrong with your site. Have you made modifications to the plugins files?

**Comment by technologynhaa** [comment]
This is the strange part.  Importing meetingsdata.csv works perfectly even
though I recognize that it is not in the right format.  So, I do not want
to create tsml_import_reformat, I want to find it or whatever is taking
meetingsdata.csv and making it usable.  Ideally, I would like Area
Fellowship to support your format.  To do this I need to take out any
custom import reformatting code.

**Comment by technologynhaa** [comment]
OK...I found the code.  It is located in X-child, just as you indicated and
is below.  So, in theory, if I comment everything out, importing
meetings.csv will work.  The next step will be to get Area Fellowship to
output meetingsdata.csv natively with the proper District data and then we
are good to go.  Possibly easier said than done, but at least we are closer
to understanding the mystery :-)

Yours in Service,

Sean K.
Area 43 Technology Chair

<?php
//
=============================================================================
// FUNCTIONS.PHP
//
-----------------------------------------------------------------------------
// Overwrite or add your own custom functions to X in this file.
//
=============================================================================
//
=============================================================================
// TABLE OF CONTENTS
//
-----------------------------------------------------------------------------
//   01. Enqueue Parent Stylesheet
//   02. Additional Functions
//
=============================================================================
// Enqueue Parent Stylesheet
//
=============================================================================
add_filter( 'x_enqueue_parent_stylesheet', '__return_true' );
// Additional Functions
//
=============================================================================


// 12 Step Meeting List Code
//
-----------------------------------------------------------------------------

function strip_parens($string) {
$count = 0;
$chars = str_split($string);
$return = '';
foreach($chars as $char) {
if ($char == '(') {
$count++;
} elseif ($char == ')') {
$count--;
} elseif ($count == 0) {
$return .= $char;
}
}
return trim(preg_replace('/\s+/', ' ', $return));
}

tsml_custom_types(array(
'X' => 'Wheelchair Access',
'SP' => 'Speaker',
'D' => 'Discussion',
'C' => 'Closed',
'O' => 'Open',
'ST' => 'Step Meeting',
'B' => 'Big Book Meeting',
'ASL' => 'Hearing Impaired',
'FR' => 'French Speaking',
'S' => 'Spanish Speaking',
'LGBTQ' => 'LGBTQ',
'ONL' => 'Online',
'TC' => 'Temporary Closure',
'AL-AN' => 'Al-Anon at Same Time/Place',
));

function tsml_import_reformat($input) {

//meeting types lookup
$types = array(
'X' => 'Wheelchair Access',
'S' => 'Speaker',
'D' => 'Discussion',
'C' => 'Closed',
'O' => 'Open',
'T' => 'Step Meeting',
'B' => 'Big Book Meeting',
'+' => 'Hearing Impaired',
'F' => 'French Speaking',
'&' => 'Spanish Speaking',
'Q' => 'LGBTQ',
'ONL' => 'Online',
'TC' => 'Temporary Closure',
'*' => 'Al-Anon at Same Time/Place',
);

//start meetings array with header row
$meetings = array(array(
'day',
'region',
'time',
'end time',
'city',
'state',
'country',
'name',
'address',
'location',
'types',
'notes',
'Conference URL',
'Conference Phone'
));

//days lookup
$days = array(
1 => 'Sunday',
2 => 'Monday',
3 => 'Tuesday',
4 => 'Wednesday',
5 => 'Thursday',
6 => 'Friday',
7 => 'Saturday',
);

//columns of the input file we're reading in
$columns = array(
'day_description',
'day',
'start',
'end',
'city',
'name',
'location',
'address',
'types',
'region',
'gso',
'conf_url',
'conf_phone'
);

//regions lookup
$regions = array(
7 => 'District 7: Peterborough Area',
8 => 'District 8: Concord Area',
9 => 'District 9: Derry Area',
10 => 'District 10: Salem Area',
11 => 'District 11: New Hampton Area',
12 => 'District 12: Manchester Area',
13 => 'District 13: Nashua Area',
14 => 'District 14: Portsmouth Area',
15 => 'District 15: Dover-Rochester Area',
16 => 'District 16: Wolfeboro Area',
17 => 'District 17: North Country Area',
18 => 'District 18: Claremont Area',
19 => 'District 19: Keene Area',
20 => 'District 20: North Conway Area',
21 => 'District 21: Hanover Area',
);

foreach ($input as $values) {

//clean up and set up
$values = array_map('trim', $values);
$row = array_combine($columns, $values);

//error checking
if (!array_key_exists($row['day'], $days)) {
continue;
//die('day number not found. must be integer 1-7. day was ' . $row['day']);
}

//append non-standard day descripton to meeting name, eg '1st Sunday'
if (!in_array($row['day_description'], $days)) {
$row['name'] .= ' (' . $row['day_description'] . ')';
}

if ( ! array_key_exists( 'conf_url', $row ) )
$row['conf_url']="";

if ( ! array_key_exists( 'conf_phone', $row ) )
$row['conf_phone']="";

//fix types
$row['types'] = explode(',', $row['types']);
$row['types'] = array_intersect($row['types'], array_keys($types));
$row['types'] = array_map(function($input) use ($types) {
return $types[$input];
}, $row['types']);

// fix zoom ID -- registrar is using "Zoom ID: ###-###-###" instead of URL.
if ( in_array( 'Online', $row['types'] ) ) {
$url = $row['conf_url'];
if ( strpos($url,'http') === false ) {
// Not a URL - let's fix that.
$zoomid = trim( substr( $url, strpos( $url, ':' ) + 1 ) );
$zoomid = str_replace( '-', '', $zoomid );
$row['conf_url']='https://zoom.us/j/' . $zoomid;
}
}

//clean up parens from address
$cleaned_address = strip_parens($row['address']);

//notes
$notes = array();
if (array_key_exists($row['region'], $regions)) $notes[] =
$regions[$row['region']];
if (!empty($row['gso'])) $notes[] = 'GSO #' . $row['gso'];
if (strstr($row['address'], '(')) $notes[] = 'Address notes: ' .
$row['address'];

$notes = implode('<br>', $notes);

$meetings[] = array(
$days[$row['day']],
$row['city'],
$row['start'],
$row['end'],
$row['city'],
'New Hampshire',
'US',
$row['name'],
$cleaned_address,
$row['location'],
implode(',', $row['types']),
$notes,
$row['conf_url'],
$row['conf_phone']
);

}

return $meetings;
}

**Comment by tech2serve** [comment]
Good job. And that would likely explain why the import discrepancy. 

So I think we can close this?

**Comment by technologynhaa** [comment]
Thank you.  It is working in that it is importing :-). My next challenge is
to see if I can get the district filter working.  Then I will go to Area
Fellowship to see if I can get them to export the data in the proper
format.  Boy, it is not nice when people do not follow standards...Yes,
please close this ticket and I will open a new one if I need to.  Thank you
for your help!

**Comment by tech2serve** [comment]
It isn't clear to me why so many of your service entities are using databases outside of TSML if they're all then trying to import into TSML? Admittedly, it isn't important that I understand that :-) But it might be something for you all to discuss. The simplest setup is to get the data into TSML once, and then manage it there.

Websites running TSML can feed each other if you set it up that way. You do need to manually refresh (we're working on automating this, but not there yet), but then only your districts need to manage their meetings, and they feed the Area site. Area can then feed Meeting Guide, or each district can (although that is kind of unnecessary). 

Personally, that is what I would strive for. If groups must keep data in a different database (usually because they're tracking more than just meeting information), then I agree with trying to get them to export to the standard used by TSML for the import. As you learned, while creating a child theme is powerful and useful for local customizations, going so far as to implement a `tsml_reformat_function` can lead to confusion for others. I admit, you're the first area I know of that is using that feature. 

Anyway, good luck. If you end up needing help with the php side of writing or updating custom formatters, the `spec` repo here does include a number of examples for how to pull data from a database and to reformat it to a proper Meeting Guide feed. In your case, you would need to add writing that to a csv, but you could also experiment with setting it up as a JSON feed to the district site in question. Lots of possibilities.

**Comment by technologynhaa** [comment]
As I understand the workflow, groups submit information to our registrar
who then inserts it into Area Fellowship.  This will eventually be whatever
GSO uses, but as far as I know the ERP deployment there has been
challenging.  As you bring up, the issue is that the Area Registrar handles
far more information than just meeting data (e.g. officers, group
information, etc.). This all predates me, but my goal is to confirm to TSML
standards as much as possible in order to be able to stay up with your
updates and support maintainability.  We do have a separate issue with
district to Area synchronization and one of the members of my technology
committee is looking at that.  I would also prefer that we use the JSON
feed, but we need to crawl before we can walk, I think (unless Area
Fellowship says jumping to JSON is easier).
